keps/0004-add-testing-infrastructure.md

```diff
@@ -68,21 +80,48 @@ This is a two phase proposal.
 
 ### Tests
 
-Add testing packages to:
+KUDO will be tested by several different test suites that validate individual components, KUDO, and frameworks.
+
+#### Unit tests
+
+Unit tests test a single component and require no network connectivity. These should be fast, run on every pull request,
+and test that individual KUDO components behave correctly. Unit tests are written as standard Go tests.
+
+All packages should incorporate unit tests that validate their behavior. Some examples of tests that should be included are:
 
 - pkg/apis - Conformance of objects, serialization, validation
 - pkg/controller - Controllers implement the correct process flows for changes in Instance objects, and creation of PlanExecutions
 - pkg/util - Util functions behave as expected
 
-Integration tests
+Go coverage reports will be used to ensure that changes do not reduce test coverage.
+
+#### Integration tests
+
+Integration tests test the KUDO controller in its entirety. These tests can be run against either a real Kubernetes cluster
+or a local control plane. Initially, all integration tests will be run against every pull request. If this begins to degrade
+velocity, integration tests can be run on master or nightly builds. Integration tests can be written either in Go or using
+the test harness designed in KEP-0008.
```

Comment:
I think it would be good to have this on all PRs, but maybe not on every commit:

After a successful review, an approval can be given, at whcih point the integration suite could be run against the PR before final merge:

https://circleci.com/docs/2.0/triggers/#manual-approval
https://discuss.circleci.com/t/tag-label-filter/11158

Reply:
Ah, good idea!

Reply:
> Integration tests can be written either in Go or using the test harness designed in KEP-0008.

This choice makes a bit ambiguous. 

1. It might be a good idea to find if the test-harness is suitable for writing any integration test, and recommend that any integration test will be written using the test harness designed in KEP-0008.

2. Or Recommend that integration test will be written in Go and And test harness is utilized only for framework tests.

A stronger recommendation will be desirable, and ideally a suggestion to use a single test framework (/harness),

Reply:
I agree, though I also didn't want to preclude being able to write tests in Go in case something comes up that cannot be tested using the harness - but maybe better to suggest extending the harness if that happens.